### PR TITLE
Added flag -s-template-id to use another segmentation (e.g. white matter) for registration to template

### LINKED ISSUE #3297: Make it possible to use another segmentation, e.g. white matter, for registration
### Context

Currently, the function assumes that the user inputs the cord segmentation with the flag `-s`. SCT then automatically associates `-s` with the file with ID=3 from **info_label.txt** file:

https://github.com/neuropoly/spinalcordtoolbox/blob/6d70ad82fbd9ba7961fc1ef7fbb7ebca70ac6fc7/spinalcordtoolbox/scripts/sct_register_to_template.py#L342

This ID label is supposed to correspond to the spinal cord mask.

However, in some cases where the white matter mask is available, the user might want to use it for registration, as it would provide better registration performance.

In this case, we need to specify somehow that another segmentation ID file should be used.

### Suggestion

Introduce the flag `-s-template-id`, which would indicate which file to select to be associated with `-s`. By default, the value would be "3" to ensure retro-compatibility.

### PR DESCRIPTION
**Pull request by jcohenadad**
#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've applied a [release milestone](https://github.com/neuropoly/spinalcordtoolbox/milestones) (major, minor, patch) in line with [Semantic Versioning guidelines](https://github.com/neuropoly/spinalcordtoolbox/wiki/Misc%3A-Creating-a-new-release#convention-for-naming-releases) 
- [x] I've assigned a reviewer

#### PR contents

- [x] I've consulted [SCT's internal developer documentation](https://github.com/neuropoly/spinalcordtoolbox/wiki) to ensure my contribution is in line with any relevant design decisions
- [ ] I've added [relevant tests](https://github.com/neuropoly/spinalcordtoolbox/wiki/Programming%3A-Tests) for my contribution
- [ ] I've updated the [relevant documentation](https://github.com/neuropoly/spinalcordtoolbox/wiki/Programming%3A-Documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description

Currently, the function assumes that the user inputs the cord segmentation with the flag `-s`. SCT then automatically associates `-s` with the file with ID=3 from **info_label.txt** file:

https://github.com/neuropoly/spinalcordtoolbox/blob/6d70ad82fbd9ba7961fc1ef7fbb7ebca70ac6fc7/spinalcordtoolbox/scripts/sct_register_to_template.py#L342

This ID label is supposed to correspond to the spinal cord mask.

However, in some cases where the white matter mask is available, the user might want to use it for registration, as it would provide better registration performance.

In this case, we need to specify somehow that another segmentation ID file should be used.

In this PR we introduce the flag `-s-template-id`, which indicates which file to select to be associated with `-s`. By default, the value would be "3" to ensure retro-compatibility.

## Linked issues
Fixes #3297

### CODE REVIEW

File: spinalcordtoolbox/scripts/sct_register_to_template.py [hunk 1]
```diff
@@ -135,6 +135,15 @@ def get_parser():
         action="help",
         help="Show this help message and exit."
     )
+    optional.add_argument(
+        '-s-template-id',
+        metavar=Metavar.int,
+        type=int,
+        help="Segmentation from the template to associate with the flag '-s'. Typically, the spinal cord segmentation "
+             "is used, but if it is available, the white matter mask would produced better registration results. The ID"
```
Comment (kousu) [inline spinalcordtoolbox/scripts/sct_register_to_template.py:143]: Is this "would" or "could"?

File: spinalcordtoolbox/registration/landmarks.py [hunk 2]
```diff
@@ -288,14 +284,14 @@ def getRigidTransformFromLandmarks(points_dest, points_src, constraints='Tx_Ty_T
...
         # plt.show()
-        plt.savefig(os.path.join(path_qc, 'getRigidTransformFromLandmarks_plot.png'))
+        plt.savefig('getRigidTransformFromLandmarks_plot.png')
 
         fig2 = plt.figure()
         plt.plot(sse_results)
         plt.grid()
         plt.title('#Iterations: ' + str(res.nit) + ', #FuncEval: ' + str(res.nfev) + ', Error: ' + str(res.fun))
         plt.show()
-        plt.savefig(os.path.join(path_qc, 'getRigidTransformFromLandmarks_iterations.png'))
```
Comment (kousu) [inline spinalcordtoolbox/registration/landmarks.py:298]: Cool. I like relying on `pwd` much better. That seems nicely unixey.

File: spinalcordtoolbox/scripts/sct_register_to_template.py [hunk 3]
```diff
@@ -135,6 +135,15 @@ def get_parser():
...
+             "is indicated in the file 'template/info_label.txt' of the template indicated by the flag '-t'.",
+        default=3
```
Comment (kousu) [inline spinalcordtoolbox/scripts/sct_register_to_template.py:145]: I think this is probably going to drive people crazy. It's a double-layer of indirection, via an integer. Could it at least be an enum instead of an int?

But, looking at the only template people actually use, `data/PAM50/template/` (to remind myself and the future: this only shows up after `./install_sct`; it's source is https://github.com/sct-data/PAM50/releases/download/r20201104/PAM50-r20201104.zip), the options are:


```
# Template labels
# Keyword=IndivLabels (Please DO NOT change this line)
# ID, name, file
0, T1-weighted template, PAM50_t1.nii.gz
1, T2-weighted template, PAM50_t2.nii.gz
2, T2S-weighted template, PAM50_t2s.nii.gz
3, spinal cord mask (binary), PAM50_cord.nii.gz
4, white matter mask (probabilistic), PAM50_wm.nii.gz
5, gray matter mask (probabilistic), PAM50_gm.nii.gz
6, cerebrospinal fluid mask (binary), PAM50_csf.nii.gz
7, spinal cord mask with discrete vertebral levels, PAM50_levels.nii.gz
8, spinal cord mask with continuous vertebral levels, PAM50_levels_continuous.nii.gz
9, point-wise vertebral body labels, PAM50_label_body.nii.gz
10, point-wise intervertebral disc labels, PAM50_label_disc.nii.gz
11, point-wise posterior intervertebral disc labels, PAM50_label_discPosterior.nii.gz
12, spine mask (binary), PAM50_spine.nii.gz
13, spinal cord centerline, PAM50_centerline.nii.gz
14, point-wise spinal level labels, PAM50_label_spinal_levels.nii.gz
```

So, `info_label.txt` is a manifest, or maybe a dictionary. And it already has unique IDs. So instead of taking an int here, what if you took the string "cord", or "wm" or "gm". That way the scripts people run and what they see on disk match up more directly. You can keep `info_label.txt` for the full names, but for the future I'd also think about dropping its first column and swapping the third and second around so it looks like like the key-value dictionary it is.

You'll have to rework (or maybe just drop?) `get_file_label()`, then.

As written, especially with this explanation, it looks like *every* template has the same set of options. But that's not true, is it? If that was true, why ship `info_label.txt` with each?

**Review by kousu — COMMENTED** [review]
Is there an example I could see of actually running this? Is there something in one of the SCT courses?

I'm missing a lot of context for understanding what selecting different segmentations should do. I don't think I would know what to look for.

File: spinalcordtoolbox/scripts/sct_register_to_template.py (hunk 1 above)
Comment (jcohenadad) [reply spinalcordtoolbox/scripts/sct_register_to_template.py:143]: good catch-- could is more accurate-- can you pls update the code-- thx

File: spinalcordtoolbox/scripts/sct_register_to_template.py (hunk 3 above)
Comment (jcohenadad) [reply spinalcordtoolbox/scripts/sct_register_to_template.py:145]: @kousu these are excellent thoughts-- quick context (no time to elaborate more): a few years ago we did use more intuitive naming (sc, wm, etc.) however for reasons i vaguely remember (we need to dig in the past PRs, i don't have time to do that now) this caused some issues because as we were prototyping register to template functionality and adding more contrast (t1, t2s, etc.) it because cumbersome to use naming (eg: there would be a "label" for "vertebral label" and then we introduced spinal label, so "label" was no more specific enough, etc.), so we ended up using numerical IDs. However, i've been unhappy about this choice for the past few years-- i'd like to come back to your suggestion-- if you or someone could please open an issue to get this started that would be great-- for now, this PR needs go get in because users are waiting after us-- they have urgent analyses to perform-- thanks

**Comment by jcohenadad** [comment]
> Is there an example I could see of actually running this? Is there something in one of the SCT courses?
> 
> I'm missing a lot of context for understanding what selecting different segmentations should do. I don't think I would know what to look for.

i'm really sorry but i don't have the time to provide this guidance now-- the gist of it is: selecting the wm mask (instead of the cord mask) would include the cord contour AND the gm contour, so that registration to template can rely on the cord contour AND gm contour-- the typical usage of reg to template ONLY uses the cord contour-- tehre is a mentino on that in the batch processing-- it is currently commented-- @joshuacwnewton opened an issue to actually update this commented part-- but it relates to this--

File: spinalcordtoolbox/scripts/sct_register_to_template.py [hunk 4]
```diff
@@ -135,6 +135,15 @@ def get_parser():
         action="help",
         help="Show this help message and exit."
     )
+    optional.add_argument(
+        '-s-template-id',
+        metavar=Metavar.int,
+        type=int,
+        help="Segmentation from the template to associate with the flag '-s'. Typically, the spinal cord segmentation "
+             "is used, but if it is available, the white matter mask would produced better registration results. The ID"
+             "is indicated in the file 'template/info_label.txt' of the template indicated by the flag '-t'.",
```
Comment (kousu) [inline spinalcordtoolbox/scripts/sct_register_to_template.py:144]:
```suggestion
        help="Segmentation from the template to be used by flag '-s'. By default, the spinal cord segmentation "
             "is used, but if available, a different segmentation could produced better registration results. The ID"
             "is an integer indicated in the file 'template/info_label.txt' of the template indicated by the flag '-t'.",
```

Comment (jcohenadad) [reply spinalcordtoolbox/scripts/sct_register_to_template.py]: no that's not it. "-s" is the seg from the subject, while this new flag "-s-template-id" is the seg from the template. that's why i used the term "associate with"-- could you suggest something clearer maybe? if not we stick with associate

**Comment by joshuacwnewton** [comment]
Just commenting to mention that I'd like to review this today/tomorrow. (A personal issue has come up, which is why I haven't reviewed this yet, but I did see it and skim over it.)

**Review by kousu — COMMENTED** [review]
I want to help you out but I have realized I don't feel qualified to to test it. I don't know how what datasets I would use or what to expect results to look like.

I'm going to leave this with @joshuacwnewton, who reminded me that this should be a feature bump: 5.2.1 -> 5.3.0.

**Comment by kousu** [comment]
I did appease the linter for you though so you don't need to worry about that.

File: spinalcordtoolbox/scripts/sct_register_to_template.py (hunk 3 above)
Comment (kousu) [reply spinalcordtoolbox/scripts/sct_register_to_template.py:146]: #3303 

I also edited your explanation so that it doesn't talk about white matter, because each template has many or few segmentations and there's no reason any of them have to be white matter or that the user would want to use that one. Unless this is somewhere where my knowledge is failing me.

Comment (jcohenadad) [reply spinalcordtoolbox/scripts/sct_register_to_template.py:146]: thank you @kousu, actually i think i will put back WM, because in 99% of use cases, ppl have the cord seg, in 20% of the cases they have cord and WM, and in <1% of the cases they have another seg (eg: MS lesion, spine seg, etc.).

Comment (jcohenadad) [reply spinalcordtoolbox/scripts/sct_register_to_template.py:146]: thank you @kousu, actually i think i will put back WM, because in 99% of use cases, ppl have the cord seg, in 20% of the cases they have cord and WM, and in <1% of the cases they have another seg (eg: MS lesion, spine seg, etc.).

File: spinalcordtoolbox/registration/landmarks.py [hunk 5]
```diff
@@ -288,14 +286,14 @@ def getRigidTransformFromLandmarks(points_dest, points_src, constraints='Tx_Ty_T
...
+        plt.savefig(os.path.join('getRigidTransformFromLandmarks_iterations.png'))
```
Comment (joshuacwnewton) [inline spinalcordtoolbox/registration/landmarks.py:296]: Nit: There is an extra `os.path.join` here.

File: spinalcordtoolbox/registration/landmarks.py [hunk 6]
```diff
@@ -41,7 +41,8 @@
 ini_param_trans_y = -150.0  # pix
 initial_step = 2
 
-def register_landmarks(fname_src, fname_dest, dof, fname_affine='affine.txt', verbose=1, path_qc=None):
+
+def register_landmarks(fname_src, fname_dest, dof, fname_affine='affine.txt', verbose=1):
```
Comment (joshuacwnewton) [inline spinalcordtoolbox/registration/landmarks.py:45]: What is the motivation for the changes in `landmarks.py`? 

(I'm not necessarily opposed, they are just unexpected to me given the scope of this PR. I am wondering why they were included when they appear to be unrelated.)

File: spinalcordtoolbox/registration/landmarks.py [hunk 7]
```diff
@@ -101,6 +102,7 @@ def register_landmarks(fname_src, fname_dest, dof, fname_affine='affine.txt', ve
                                                            points_moving_barycenter[2]))
     text_file.close()
 
+
```
Comment (joshuacwnewton) [inline spinalcordtoolbox/registration/landmarks.py:105]: Small nit (see Lintly feedback)

**Review by joshuacwnewton — APPROVED** [review]
I don't have much to say besides what Nick has already pointed (int labels vs. descriptive labels). 

These points are straightfoward, so I'll pre-approve. :slightly_smiling_face:

File: spinalcordtoolbox/registration/landmarks.py (hunk 5 above)
Comment (jcohenadad) [reply spinalcordtoolbox/registration/landmarks.py:296]: sorry no time to wait for new ci--

File: spinalcordtoolbox/registration/landmarks.py (hunk 6 above)
Comment (jcohenadad) [reply spinalcordtoolbox/registration/landmarks.py:45]: > is it just because path_qc was unused by calls to register_landmarks?

indeed that's it-- indeed, out of scope, i just did it to save time

**Comment by jcohenadad** [comment]
thank you so much for your review @joshuacwnewton @kousu
